File: fact.py
#!/usr/bin/env python
# -*- coding: utf8 -*-

# @@@ get_version: python -V 2>&1 | awk '{print $2}'

times = 2000000
fact_16 = 20922789888000

def fact(n):
    if n == 1:
        return 1
    return n * fact(n - 1)

ok = True
for i in xrange(times):
    ok = ok and fact(16) == fact_16

print "python finish %i - %s" % (times, 'ok' if ok else 'fail')

File: fact.hs
-- @@@ get_version: ghc --version | awk '{print $8}'
-- @@@ before: ghc -o fact_hs fact.hs
-- @@@ instead: ./fact_hs
-- @@@ after: rm fact_hs fact.o fact.hi

module Main where

times = 10000000
fact_16 = 20922789888000

factorial :: Integer -> Integer
factorial 0 = 1
factorial n = n * factorial(n - 1)

main = do
    let result = if foldr1 (&&) [factorial(16) == 20922789888000 | x <- takeWhile (<= times) [1..]]
                 then "ok"
                 else "fail"
    putStrLn("Haskell finish " ++ (show times) ++ " - " ++ result)
